In [1]:
import pandas as pd
import glob
import ntpath

In [2]:
# Get CSV files list from a folder
path = 'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/metadata'
csv_files = glob.glob(path + "/*.csv")

# Read each CSV file into DataFrame
# This creates a list of dataframes
df_list = (pd.read_csv(file,sep = ',') for file in csv_files)

# Concatenate all DataFrames
metadata = pd.concat(df_list, ignore_index=True)
metadata.rename(columns={'author':'main_author'}, inplace=True)
metadata

id  review_number  revision_number    main_author     status  \
0       12027_rev1          12027                1  Emerson Nolan  ABANDONED   
1       12050_rev1          12050                1  Emerson Nolan     MERGED   
2       12050_rev2          12050                2  Emerson Nolan     MERGED   
3       12050_rev3          12050                3  Emerson Nolan     MERGED   
4       12050_rev4          12050                4  Emerson Nolan     MERGED   
...            ...            ...              ...            ...        ...   
144901  85939_rev1          85939                1   Johan Savage  ABANDONED   
144902  85948_rev1          85948                1   Emely Wilson     MERGED   
144903  85948_rev2          85948                2   Emely Wilson     MERGED   
144904  85957_rev1          85957                1   Emely Wilson     MERGED   
144905  85957_rev2          85957                2   Emely Wilson     MERGED   

                                        change_id  \
0       I634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       Idd6454bd6574a9337528281df912f51a9fda52ad   
2       Idd6454bd6574a9337528281df912f51a9fda52ad   
3       Idd6454bd6574a9337528281df912f51a9fda52ad   
4       Idd6454bd6574a9337528281df912f51a9fda52ad   
...                                           ...   
144901  I116a423fc451b41fa324ddae18a9703f9b96c4df   
144902  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144903  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144904  I6fdb6c92a74cda225a1ae82b732111754a27735b   
144905  I6fdb6c92a74cda225a1ae82b732111754a27735b   

                                            url  \
0       http://review.couchbase.org/#/c/12027/1   
1       http://review.couchbase.org/#/c/12050/1   
2       http://review.couchbase.org/#/c/12050/2   
3       http://review.couchbase.org/#/c/12050/3   
4       http://review.couchbase.org/#/c/12050/4   
...                                         ...   
144901  http://review.couchbase.org/#/c/85939/1   
144902  http://review.couchbase.org/#/c/85948/1   
144903  http://review.couchbase.org/#/c/85948/2   
144904  http://review.couchbase.org/#/c/85957/1   
144905  http://review.couchbase.org/#/c/85957/2   

                       original_before_commit_id  \
0       b73b6c8f30dd87337f9fb68349bfdace565c48c8   
1       31f14b3b541cf1322f76683ff93947389a92f39b   
2       31f14b3b541cf1322f76683ff93947389a92f39b   
3       627edb710baa868773d5f09e6ee1842ceff34fe8   
4       627edb710baa868773d5f09e6ee1842ceff34fe8   
...                                          ...   
144901  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144902  8585de657e924878715ebfc49ce61be290fbde6f   
144903  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144904  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144905  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   

                        original_after_commit_id  \
0       634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       dd6454bd6574a9337528281df912f51a9fda52ad   
2       8c338944a8c4e5e068db7009c41be5cbcb8b72f1   
3       eddbfeac3b56721994ab0bdc7a63ded3aec6bb1a   
4       ced15bea74bce3e242865edd18d5cb1370d2eecb   
...                                          ...   
144901  c9d7b2eaaf452e75925255a77f265c049b5c5b57   
144902  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144903  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   
144904  f46758df755d9ad5d510add3f128ad7a83ded20b   
144905  0959b5a83b848343dced887b034e49308c53097e   

                                before_commit_id  \
0       2a7010a1c308cf0e635fdc5fc2bb5f50890fe9c6   
1       2c9169e853fd0dafc3a4a7779cd557247fc0e1cf   
2       b28453bc3192cd4f5ef3c5ba60a6cef08dbb3e86   
3       5e304c853e6ab952cc609ef3b002b1d5f91b7ca6   
4       6cbf9ff09a6ec466c1fca2679d25b7293f3e41c8   
...                                          ...   
144901  084583792c84d761b8fc2980be7c080ec92f7ff5   
144902  98cc475c634931f7dd59c50614548ddd2012f445   
144903  b0645bbe00d8d82d55069b3d8352409bf7f755bf   
144904  4c982463a76ebdaee7c26fdbc5d

Add _discussion.txt in the end of Pull request number in order to compare them with all_discussion_specs dataframe from below.

In [3]:
metadata['id'] = metadata['id'].astype(str) + '_discussion.txt'

Pr_number column will be used for merge reasons.

In [4]:
metadata['PR_number'] = metadata['id'].astype(str).str[:10]

Drop files with same name

In [5]:
metadata = metadata.drop_duplicates(subset=['PR_number'])

In [6]:
metadata

id  review_number  revision_number  \
0       12027_rev1_discussion.txt          12027                1   
1       12050_rev1_discussion.txt          12050                1   
2       12050_rev2_discussion.txt          12050                2   
3       12050_rev3_discussion.txt          12050                3   
4       12050_rev4_discussion.txt          12050                4   
...                           ...            ...              ...   
144901  85939_rev1_discussion.txt          85939                1   
144902  85948_rev1_discussion.txt          85948                1   
144903  85948_rev2_discussion.txt          85948                2   
144904  85957_rev1_discussion.txt          85957                1   
144905  85957_rev2_discussion.txt          85957                2   

          main_author     status                                  change_id  \
0       Emerson Nolan  ABANDONED  I634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
2       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
3       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
4       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
...               ...        ...                                        ...   
144901   Johan Savage  ABANDONED  I116a423fc451b41fa324ddae18a9703f9b96c4df   
144902   Emely Wilson     MERGED  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144903   Emely Wilson     MERGED  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144904   Emely Wilson     MERGED  I6fdb6c92a74cda225a1ae82b732111754a27735b   
144905   Emely Wilson     MERGED  I6fdb6c92a74cda225a1ae82b732111754a27735b   

                                            url  \
0       http://review.couchbase.org/#/c/12027/1   
1       http://review.couchbase.org/#/c/12050/1   
2       http://review.couchbase.org/#/c/12050/2   
3       http://review.couchbase.org/#/c/12050/3   
4       http://review.couchbase.org/#/c/12050/4   
...                                         ...   
144901  http://review.couchbase.org/#/c/85939/1   
144902  http://review.couchbase.org/#/c/85948/1   
144903  http://review.couchbase.org/#/c/85948/2   
144904  http://review.couchbase.org/#/c/85957/1   
144905  http://review.couchbase.org/#/c/85957/2   

                       original_before_commit_id  \
0       b73b6c8f30dd87337f9fb68349bfdace565c48c8   
1       31f14b3b541cf1322f76683ff93947389a92f39b   
2       31f14b3b541cf1322f76683ff93947389a92f39b   
3       627edb710baa868773d5f09e6ee1842ceff34fe8   
4       627edb710baa868773d5f09e6ee1842ceff34fe8   
...                                          ...   
144901  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144902  8585de657e924878715ebfc49ce61be290fbde6f   
144903  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144904  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144905  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   

                        original_after_commit_id  \
0       634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       dd6454bd6574a9337528281df912f51a9fda52ad   
2       8c338944a8c4e5e068db7009c41be5cbcb8b72f1   
3       eddbfeac3b56721994ab0bdc7a63ded3aec6bb1a   
4       ced15bea74bce3e242865edd18d5cb1370d2eecb   
...                                          ...   
144901  c9d7b2eaaf452e75925255a77f265c049b5c5b57   
144902  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144903  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   
144904  f46758df755d9ad5d510add3f128ad7a83ded20b   
144905  0959b5a83b848343dced887b034e49308c53097e   

                                before_commit_id  \
0       2a7010a1c308cf0e635fdc5fc2bb5f50890fe9c6   
1       2c9169e853fd0dafc3a4a7779cd557247fc0e1cf   
2       b28453bc3192cd4f5ef3c5ba60a6cef08dbb3e86   
3       5e304c853e6ab952cc609ef3b002b1d5f91b7ca6   
4       6cbf9ff09a6ec466c1fca2679d25b7293f3e41c8   
...                                          ...   
144901  084583792c84d761b8

For discussion data.

In [7]:
def add_value(dict_obj, key, value):
    ''' Adds a key-value pair to the dictionary.
        If the key already exists in the dictionary, 
        it will associate multiple values with that 
        key instead of overwritting its value'''
    if key not in dict_obj:
        dict_obj[key] = value
    elif isinstance(dict_obj[key], list):
        dict_obj[key].append(value)
    else:
        dict_obj[key] = [dict_obj[key], value]

Find and delete all files with duplicate names

In [8]:
files_path = glob.glob('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/*/*/*.txt')

In [9]:
file_names = []
duplicates = []
for file in files_path:
    if(ntpath.basename(file) in file_names):
        duplicates.append(ntpath.basename(file))
    file_names.append(ntpath.basename(file))

In [10]:
print(len(duplicates))

11879


In [11]:
for file in files_path:
    if(ntpath.basename(file) in duplicates):
        os.remove(file)

In [12]:
files_path = glob.glob('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/*/*/*.txt')

In [37]:
print(len(wrong))

0


Search if there are duplicates

In [35]:
duplicates_files=[]
wrong=[]
for i in files_path:
    if ntpath.basename(i) not in duplicates_files:
        duplicates_files.append(ntpath.basename(i))
    else:
        wrong.append(ntpath.basename(i))

In [38]:
print(len(wrong))

0


In [39]:
print(len(duplicates_files))

129483


Find all commit messages, descriptions, comments. their author and date.

In [14]:
files_path = glob.glob('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/*/*/*.txt')
commit_messages={}
descriptions={}
all_comments = {}
all_authors = {}
all_dates = {}

for file_path in files_path:
    comments = {}
    comments_limits=[]
    authors = {}
    dates = {}
    file = open(file_path, encoding='utf8')
# Checks for commit messages
    for x, line in enumerate(file):
        if line[:12] == "First commit":
            commit_message_start = x
        if line[:10] == "Change-Id:":
            commit_message_end = x
            break
    file.close()
# Checks for description
    file = open(file_path, encoding='utf8')
    for x, line in enumerate(file):
# If it has a commit message
        if line[:12] == "First commit":
            description_end = x
            break
        if line[:10] == "Change-Id:":
            description_end = x
            break
    file.close()
    
    # Check for Comments
    file = open(file_path, encoding='utf8')
    for x, line in enumerate(file):
        if line[:5] == "date:":
            start = x
        if line[:10] == "----------":
            end = x
            tup=(start,end)
            comments_limits.append(tup)
    file.close()
    
    comment = ""
    author = ""
    date = ""
    commit_message = ""
    description = ""
    # Saves all commit messages and descriptions
    file = open(file_path, encoding='utf8')
    for x, line in enumerate(file):
        if x > commit_message_start and x < commit_message_end:
            commit_message+=line
        if x > 1 and x < description_end:
            description+=line
    add_value(commit_messages, ntpath.basename(file_path)[:10], commit_message.replace("\n",""))
    add_value(descriptions, ntpath.basename(file_path)[:10], description.replace("\n",""))
    file.close()
    
    # Saving the comments, author and dates of the comments
    for tuple in comments_limits:
        file = open(file_path, encoding='utf8')
        for x, line in enumerate(file):
            if x > tuple[0] and x < tuple[1]:
                comment+=line
            if x == tuple[0]-1:
                author+=line[8:]
            if x == tuple[0]:
                date+=line[6:]
        add_value(comments, ntpath.basename(file_path)[:10], comment.replace("\n",""))
        add_value(authors, ntpath.basename(file_path)[:10], author.replace("\n",""))
        add_value(dates, ntpath.basename(file_path)[:10], date.replace("\n",""))
        comment = ""
        author = ""
        date = ""
        file.close()
    all_comments.update(comments)
    all_authors.update(authors)
    all_dates.update(dates)

    file.close()

There are 141362 revision files (.txt) in discussion

In [59]:
len(descriptions)

119603

So we have 119603 revisions with at least one comment

Convert lists to dataframes

In [60]:
all_comments_df = pd.Series(all_comments, name='Comment').rename_axis('PR_number').explode().reset_index()
all_authors_df = pd.Series(all_authors, name='Author').rename_axis('PR_number').explode().reset_index()
all_dates_df = pd.Series(all_dates, name='Date').rename_axis('PR_number').explode().reset_index()
commit_messages_df = pd.Series(commit_messages, name='Commit_message').rename_axis('PR_number').explode().reset_index()
descriptions_df = pd.Series(descriptions, name='Description').rename_axis('PR_number').explode().reset_index()

There are 3.12 Comments per revision.

In [61]:
len(all_comments_df)/len(descriptions)

3.3536616974490605

Merge all dataframes to one

In [62]:
all_comments_df['Author'] = all_authors_df['Author'].values
all_comments_df['Date'] = all_dates_df['Date'].values

We use right as how because the commit_messages_df dataframe contains all the revisions (128607).

In [63]:
all_comments_df

PR_number                                            Comment  \
0       11287_rev2                              Uploaded patch set 2.   
1       11287_rev3                              Uploaded patch set 3.   
2       11287_rev3  Patch Set 3: Looks good to me, approvedI'm sur...   
3       11287_rev3                              Patch Set 3: Verified   
4       11287_rev3  Change has been successfully merged into the g...   
...            ...                                                ...   
401103  9977_rev3_  Patch Set 3: Looks good to me, approvedRubbers...   
401104  9977_rev3_  Change has been successfully cherry-picked as ...   
401105  9988_rev2_                              Uploaded patch set 2.   
401106  9988_rev2_  Patch Set 2: Verified; Looks good to me, approved   
401107  9988_rev2_  Change has been successfully cherry-picked as ...   

               Author                           Date  
0        Jeffrey Roth  2012-01-03 22:42:46.755000000  
1        Jeffrey Roth  2012-01-04 00:15:52.130000000  
2       Kenny Maynard  2012-01-04 00:23:41.182000000  
3        Jeffrey Roth  2012-01-04 00:49:35.390000000  
4        Jeffrey Roth  2012-01-04 00:49:35.587000000  
...               ...                            ...  
401103    Reid Stokes  2011-10-08 01:42:12.137000000  
401104    Reid Stokes  2011-10-08 01:42:16.091000000  
401105    Ayden Rivas  2011-10-06 22:47:10.792000000  
401106   Zackery Barr  2011-10-06 22:48:26.810000000  
401107   Zackery Barr  2011-10-06 22:48:28.498000000  

[401108 rows x 4 columns]

In [64]:
print(len(all_comments_df['PR_number'].unique()))

117635


In [65]:
print(len(all_discussion_specs['PR_number'].unique()))

117635


In [66]:
all_data = pd.merge(all_discussion_specs, metadata[['PR_number','status','main_author','revision_number','url']], how="left", on='PR_number')
all_data

PR_number                                            Comment  \
0       11287_rev2                              Uploaded patch set 2.   
1       11287_rev3                              Uploaded patch set 3.   
2       11287_rev3  Patch Set 3: Looks good to me, approvedI'm sur...   
3       11287_rev3                              Patch Set 3: Verified   
4       11287_rev3  Change has been successfully merged into the g...   
...            ...                                                ...   
433390  9977_rev3_  Patch Set 3: Looks good to me, approvedRubbers...   
433391  9977_rev3_  Change has been successfully cherry-picked as ...   
433392  9988_rev2_                              Uploaded patch set 2.   
433393  9988_rev2_  Patch Set 2: Verified; Looks good to me, approved   
433394  9988_rev2_  Change has been successfully cherry-picked as ...   

                                           Commit_message  status  \
0       * Transferred all of the Couchbase code from s...     NaN   
1       * Transferred all of the Couchbase code from s...     NaN   
2       * Transferred all of the Couchbase code from s...     NaN   
3       * Transferred all of the Couchbase code from s...     NaN   
4       * Transferred all of the Couchbase code from s...     NaN   
...                                                   ...     ...   
433390                                                     MERGED   
433391                                                     MERGED   
433392                                                     MERGED   
433393                                                     MERGED   
433394                                                     MERGED   

        main_author  revision_number                                     url  
0               NaN              NaN                                     NaN  
1               NaN              NaN                                     NaN  
2               NaN              NaN                                     NaN  
3               NaN              NaN                                     NaN  
4               NaN              NaN                                     NaN  
...             ...              ...                                     ...  
433390  Reid Stokes              3.0  http://review.couchbase.org/#/c/9977/3  
433391  Reid Stokes              3.0  http://review.couchbase.org/#/c/9977/3  
433392  Ayden Rivas              2.0  http://review.couchbase.org/#/c/9988/2  
433393  Ayden Rivas              2.0  http://review.couchbase.org/#/c/9988/2  
433394  Ayden Rivas              2.0  http://review.couchbase.org/#/c/9988/2  

[433395 rows x 7 columns]

In [67]:
metadata = metadata.drop_duplicates(subset=['PR_number'])

In [68]:
metadata

id  review_number  revision_number  \
0       12027_rev1_discussion.txt          12027                1   
1       12050_rev1_discussion.txt          12050                1   
2       12050_rev2_discussion.txt          12050                2   
3       12050_rev3_discussion.txt          12050                3   
4       12050_rev4_discussion.txt          12050                4   
...                           ...            ...              ...   
144901  85939_rev1_discussion.txt          85939                1   
144902  85948_rev1_discussion.txt          85948                1   
144903  85948_rev2_discussion.txt          85948                2   
144904  85957_rev1_discussion.txt          85957                1   
144905  85957_rev2_discussion.txt          85957                2   

          main_author     status                                  change_id  \
0       Emerson Nolan  ABANDONED  I634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
2       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
3       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
4       Emerson Nolan     MERGED  Idd6454bd6574a9337528281df912f51a9fda52ad   
...               ...        ...                                        ...   
144901   Johan Savage  ABANDONED  I116a423fc451b41fa324ddae18a9703f9b96c4df   
144902   Emely Wilson     MERGED  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144903   Emely Wilson     MERGED  I20931d6a69996c5d569bca62bd5b7acf5469f98a   
144904   Emely Wilson     MERGED  I6fdb6c92a74cda225a1ae82b732111754a27735b   
144905   Emely Wilson     MERGED  I6fdb6c92a74cda225a1ae82b732111754a27735b   

                                            url  \
0       http://review.couchbase.org/#/c/12027/1   
1       http://review.couchbase.org/#/c/12050/1   
2       http://review.couchbase.org/#/c/12050/2   
3       http://review.couchbase.org/#/c/12050/3   
4       http://review.couchbase.org/#/c/12050/4   
...                                         ...   
144901  http://review.couchbase.org/#/c/85939/1   
144902  http://review.couchbase.org/#/c/85948/1   
144903  http://review.couchbase.org/#/c/85948/2   
144904  http://review.couchbase.org/#/c/85957/1   
144905  http://review.couchbase.org/#/c/85957/2   

                       original_before_commit_id  \
0       b73b6c8f30dd87337f9fb68349bfdace565c48c8   
1       31f14b3b541cf1322f76683ff93947389a92f39b   
2       31f14b3b541cf1322f76683ff93947389a92f39b   
3       627edb710baa868773d5f09e6ee1842ceff34fe8   
4       627edb710baa868773d5f09e6ee1842ceff34fe8   
...                                          ...   
144901  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144902  8585de657e924878715ebfc49ce61be290fbde6f   
144903  c5009ed862c18aea4e5848f703c3bab8d8c8affb   
144904  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144905  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   

                        original_after_commit_id  \
0       634fb3ea0b12e60b8d8b27d9aa3f36d5f394d83e   
1       dd6454bd6574a9337528281df912f51a9fda52ad   
2       8c338944a8c4e5e068db7009c41be5cbcb8b72f1   
3       eddbfeac3b56721994ab0bdc7a63ded3aec6bb1a   
4       ced15bea74bce3e242865edd18d5cb1370d2eecb   
...                                          ...   
144901  c9d7b2eaaf452e75925255a77f265c049b5c5b57   
144902  08e40583a97e81d8b806706fa08dbaccb24b53ec   
144903  81442a1c9d2530ab7ab2d1a87bfa04eb89ce4568   
144904  f46758df755d9ad5d510add3f128ad7a83ded20b   
144905  0959b5a83b848343dced887b034e49308c53097e   

                                before_commit_id  \
0       2a7010a1c308cf0e635fdc5fc2bb5f50890fe9c6   
1       2c9169e853fd0dafc3a4a7779cd557247fc0e1cf   
2       b28453bc3192cd4f5ef3c5ba60a6cef08dbb3e86   
3       5e304c853e6ab952cc609ef3b002b1d5f91b7ca6   
4       6cbf9ff09a6ec466c1fca2679d25b7293f3e41c8   
...                                          ...   
144901  084583792c84d761b8

In [69]:
print(len(metadata['PR_number'].unique()))

125475


In [70]:
# pd.set_option('display.max_rows', None)

In [71]:
compression_opts = dict(method='zip',
                        archive_name='all_data.csv') 
all_data.to_csv('all_data.zip', index=False, compression=compression_opts)

In [72]:
if '11287_rev1' in all_discussion_specs['PR_number'].values :
    print("\nThis value exists in Dataframe")

For a single file (Testing)

In [73]:
import ntpath
path = 'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt'
file = open('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt')
commit_messages={}
comments = {}
comments_limits=[]
authors = {}
dates = {}
descriptionn = {}
commit_messagen = {}
# Checks for commit messages
for x, line in enumerate(file):
    if line[:12] == "First commit":
         commit_message_start = x
    if line[:10] == "Change-Id:":
        commit_message_end = x
        break
file.close()
# Checks for description
file = open('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt')
for x, line in enumerate(file):
# If it has a commit message
    if line[:12] == "First commit":
        description_end = x
        break
    if line[:10] == "Change-Id:":
        description_end = x
        break
file.close()
commit_message = ""
description = ""
# Saves all commit messages and descriptions
file = open('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt')
for x, line in enumerate(file):
    if x > commit_message_start and x < commit_message_end:
        commit_message+=line
    if x > 1 and x < description_end:
        description+=line
add_value(commit_messagen, ntpath.basename(file_path)[:10], commit_message.replace("\n",""))
add_value(descriptionn, ntpath.basename(file_path)[:10], description.replace("\n",""))
file.close()
# Check for Comments
file = open('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt')
for x, line in enumerate(file):
    if line[:5] == "date:":
        start = x
    if line[:10] == "----------":
        end = x
        tup=(start,end)
        comments_limits.append(tup)
file.close()

commit_message = ""
comment = ""
author = ""
date = ""
# For each comment in the text file
for tuple in comments_limits:
    file = open('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt')
    for x, line in enumerate(file):
        if x > tuple[0] and x < tuple[1]:
            comment+=line
# Saving the author
        if x == tuple[0]-1:
            author+=line[8:]
        if x == tuple[0]:
            date+=line[6:]
    add_value(comments, ntpath.basename(path)[:10], comment.replace("\n",""))
    #comments = ({comment.replace("\n",""),ntpath.basename("r'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/11287/11287_rev3_discussion.txt'")[:10]})
    add_value(authors, ntpath.basename("r'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt'")[:10], date.replace("\n",""))
    add_value(dates, ntpath.basename("r'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt'")[:10], author.replace("\n",""))
    comment = ""
    author = ""
    date = ""
    file.close()

#commit_messages.append(commit_message.replace('\n',""))
file.close()
print(dates)
import os
#print(os.path.splitext('C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/11287/11287_rev3_discussion.txt', sep="//")[0])

print(comments)
#print(ntpath.basename("r'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/11287/11287_rev3_discussion.txt'")[:10])


FileNotFoundError: [Errno 2] No such file or directory: 'C:/Users/GNR/Desktop/data/git_repos/crop_data/data/discussion/couchbase-java-client/12149/12149_rev2_discussion.txt'